[Data Grid] Fix row, cell and header memoizations

lauri865 opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

<!-- You can use `## Changelog` to create a description for this change in the next release. -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Think I managed to pin down everything I reported in #15618 

Before:

https://github.com/user-attachments/assets/00c64c34-e056-438f-8921-6c265a00fdf2

After:

https://github.com/user-attachments/assets/9dbe1c83-f0ef-461c-9bd4-0eba67f86c36

lauri865 commented:
Now also fixed rows having to re-render on focus/tabindex change, as that responsibility already falls to the cells. This removed an extra call to getCellParams on a row level to each cell.

lauri865 commented:
✅ All tests passing – ready for review.

romgrk commented on packages/x-data-grid/src/hooks/features/columns/gridColumnsInterfaces.ts:
```diff
@@ -6,6 +6,13 @@ export enum GridPinnedColumnPosition {
   RIGHT = 'right',
 }
 
+export enum PinnedPosition {
+  NONE,
+  LEFT,
+  RIGHT,
+  VIRTUAL,
+}
+
```
There is a strong convention to prefix all the public exports with `Grid...`, e.g. `GridPinnedPosition`. If it's for internal purposes only, it should be exported from `@mui/x-data-grid/internals` instead, then it would be ok to skip the prefix.

Also this was originally local to `GridCell` so it was somewhat clear that it was for pinned columns. If it's exported globally it probably needs a "column" in there because "pinned position" can also refer to rows.

lauri865 replied:
Actually didn't mean to to export it publicly, just ended up with circular dependencies when having it in GridCell– didn't really understand the implications. Will switch it to internals rather.

lauri865 replied:
Moved it to internals and renamed.

romgrk commented on packages/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterCell.tsx:
```diff
@@ -324,15 +316,22 @@ const GridHeaderFilterCell = React.forwardRef<HTMLDivElement, GridHeaderFilterCe
 
     const isFilterActive = isApplied || hasFocus;
 
+    let style = {
+      height,
+      width,
+      ...styleProp,
+    };
+
+    const pinnedSide = rtlFlipSide(pinnedPosition, isRtl);
+    if (pinnedSide && pinnedOffset !== undefined) {
+      style = { ...style, [pinnedSide]: pinnedOffset };
+    }
```
This block of code owns `style`, we could instead assign a value to it directly rather than making another copy.

romgrk commented on packages/x-data-grid/src/components/cell/GridCell.tsx:
```diff
@@ -350,17 +332,12 @@ const GridCell = React.forwardRef<HTMLDivElement, GridCellProps>(function GridCe
 
-      cellStyle[side] = pinnedOffset;
+    const pinnedSide = rtlFlipSide(pinnedPosition, isRtl);
+    if (pinnedSide && pinnedOffset !== undefined) {
```
Could that condition be expressed in terms of `pinnedPosition` instead? And would that allow us to keep the type of `pinnedOffset: number` instead of `?: number`? And could we move the flip logic inside the if check?

lauri865 replied:
I actually did it like that initially, but it felt wrong to me. Why create a value that shouldn't exist – could end up in adding accidental inline styles for non-pinned columns that should never exist. The type of `pinnedOffset` could be constrained to only exist if `pinnedPoisition` is left/right though.

Moving the rtlFlipSide could make sense, albeit it currently also narrows the type of pinnedPosition – if pinnedSide is defined, pinnedPosition can only be left or right, so maybe not much of a change:
<img width="740" alt="Screenshot 2024-12-03 at 00 26 17" src="https://github.com/user-attachments/assets/90b93d00-95f1-474e-ba81-f30ee2f5956e">

romgrk commented on packages/x-data-grid/src/components/columnHeaders/GridColumnHeaderItem.tsx:
```diff
@@ -287,6 +279,15 @@ function GridColumnHeaderItem(props: GridColumnHeaderItemProps) {
   const label = colDef.headerName ?? colDef.field;
 
+  const style = React.useMemo(() => {
+    let styleProp = props.style;
+    const pinnedSide = rtlFlipSide(pinnedPosition, isRtl);
+    if (pinnedSide && pinnedOffset !== undefined) {
+      styleProp = { ...styleProp, [pinnedSide]: pinnedOffset };
+    }
+    return styleProp;
```
I see this block of code repeated a few times, maybe it wants to be a function?

I'm also curious why those styles are now a bit everywhere, where were they before?

lauri865 replied:
The styles were previously defined on parent level, and passed as props as unstable objects, which broke the memoizations.

romgrk replied:
I mean we could simplify it to:

```
const styleProp = attachPinnedStyle(props.style, pinnedPosition, pinnedOffset)
```

Because I see this block at least 3-4 times I think.

romgrk commented on packages/x-data-grid/src/components/containers/GridRoot.tsx:
```diff
@@ -91,4 +97,5 @@ GridRoot.propTypes = {
 } as any;
 
-export { GridRoot };
+const GridRootMemo = React.memo(GridRoot);
```
```suggestion
const MemoizedGridRoot = fastMemo(GridRoot);
```

To keep the same style as other parts of the codebase, with `fastMemo` from `@mui/x-internals`.

romgrk commented on packages/x-data-grid/src/hooks/features/columnHeaders/useGridColumnHeaders.tsx:
```diff
@@ -63,7 +66,7 @@ export interface UseGridColumnHeadersProps {
 }
 
 export interface GetHeadersParams {
-  position?: GridPinnedColumnPosition;
+  position?: PinnedColumnPosition;
   renderContext?: GridColumnsRenderContext;
   maxLastColumn?: number;
 }
```
I'm not sure about this change, it modifies the public API. The old `GridPinnedColumnPosition` has only possible pinning values, where `PinnedColumnPosition` also includes internal implementation details such as `VIRTUAL` and `NONE`. It's annoying to have both types be named the same & represent the same thing, but I don't think we should make this change.

romgrk replied:
And possibly our users can be checking with `=== 'left'` rather than `=== PinnedPosition.LEFT`. Thoughts @mui/xgrid ?

romgrk replied:
Also it makes the constant part of the public API, so it would require the `Grid` prefix.

lauri865 replied:
Where does this leak to the public API? Must have not understood it, as it wasn't my intention to change the public types. I think public interface should be maximally constrained of course, but to an outside eye, it felt like there was so much translating happening between specific implementation of pinning and PinnedPosition currently, but for any internal implementation, it may not really worth the hassle imho.

Even if you would accidentally add a completely condition for a pinnedposition that can never exist a la `position === PinnedPosition.Whatever`, it doesn't have any impact, except for extraneous code. But the likelihood of doing that is probably next to none.

When I searched the docs, GridPinnedColumnPosition exists in two places, and those should remain unchanged, no?
https://mui.com/x/api/data-grid/grid-api/#grid-api-prop-isColumnPinned

`isColumnPinned` and `pinColumn`

romgrk commented on packages/x-license/src/Watermark/Watermark.tsx:
```diff
@@ -55,3 +55,6 @@ export function Watermark(props: WatermarkProps) {
   );
 }
+
+const WatermarkMemo = React.memo(Watermark);
+export { WatermarkMemo as Watermark };
```
Same as above for `fastMemo`.

romgrk commented on packages/x-data-grid/src/components/columnSelection/GridCellCheckboxRenderer.tsx:
```diff
@@ -92,7 +92,11 @@ const GridCellCheckboxForwardRef = React.forwardRef<HTMLInputElement, GridRender
       rootProps.rowSelectionPropagation?.parents ?? false,
     );
-    const { isIndeterminate, isChecked } = useGridSelector(apiRef, checkboxPropsSelector);
+    const { isIndeterminate, isChecked } = useGridSelector(
+      apiRef,
+      checkboxPropsSelector,
+      objectShallowCompare,
+    );
```
The proper solution should be to fix the TODO comment here:

https://github.com/mui/mui-x/blob/90c1fc2e58e53fc1d54219863e8cb6dd6a5356c3/packages/x-data-grid/src/hooks/features/rowSelection/utils.ts#L49-L50

We now have support for selectors with arguments, which is what the checkbox selectors were missing.

lauri865 replied:
I might be misunderstanding the new selectors as there's no documentation and only one implementation so far in the code base, but wouldn't it still spit out a new object and trigger a new render, even if the object values stay the same?

romgrk replied:
`createSelectorMemoizedV8` would memoize its output based on its dependencies, e.g.
```js
const checkboxSelector = createSelectorMemoizedV8(
  aSelector,
  bSelector,
  cSelector,
  (a, b, c) => { return { a, b, c } /* this result object would be memoized */ }
)
```

Though I think @MBilalShafi might handle it as part of #15200, so you could ignore this.

lauri865 replied:
This part I understand, but wouldn't it only work if a selector returns a slice of the state. In this case we're dealing with derived state, and creating a new object

In this case, you have two options:
* Use a selector that returns the derived state – return value never holds a stable reference, memoization fails.
* Put the state derivation logic in the last function – last function gets triggered also when unrelated rows get selected/deselected. Last function still creates an object with unstable object reference, everything still re-renders when other rows re-render.

So, while it fixes unrelated state triggering re-renders, it still doesn't enable us to hone in on the specific row's state. Either we need to not return an object if the row is not selected, or still need objectShallowCompare to memoize it. 🤔

romgrk replied:
> In this case we're dealing with derived state, and creating a new object

Is it clear to you that the derived state (aka the new object) is memoized by `createSelectorMemoized`? So it stays stable across state updates until one of its dependency (in the example above, `aSelector`, `bSelector` and `cSelector`) changes. That function is similar to `React.useMemo` in its behavior.

lauri865 replied:
Ok, I think I understand now after looking at the source code. Args are also passed nested into all of the subselectors, in which case `gridSortedRowIdsSelector` needs to return the correct slice based on the args.

I was first under the impression that args are only passed to the last function that takes together the selector results, in which case it would be triggered by unrelated `gridSortedRowIdsSelector` changes, and we would create a new object every time.

But if args are shared naively across any nested selectors, how do you plan to avoid any conflicts? Seems like something that needs utmost caution if used widely.

romgrk commented on packages/x-data-grid/src/components/columnSelection/GridHeaderCheckbox.tsx:
```diff
@@ -36,7 +36,9 @@ const GridHeaderCheckbox = React.forwardRef<HTMLButtonElement, GridColumnHeaderP
     const classes = useUtilityClasses(ownerState);
     const tabIndexState = useGridSelector(apiRef, gridTabIndexColumnHeaderSelector);
-    const selection = useGridSelector(apiRef, gridRowSelectionStateSelector);
+    const selection = useGridSelector(apiRef, () =>
+      gridRowSelectionStateSelector(apiRef.current.state),
+    );
```
How is this different? As far as I recall, this should be selecting the same value.

lauri865 replied:
Added it in error. Removed.

romgrk commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -207,6 +204,37 @@ export const useGridVirtualScroller = () => {
   const previousContextScrollPosition = React.useRef(EMPTY_SCROLL_POSITION);
   const previousRowContext = React.useRef(EMPTY_RENDER_CONTEXT);
   const renderContext = useGridSelector(apiRef, gridRenderContextSelector);
+  const focusedVirtualCell = useGridSelector(
+    apiRef,
+    () => {
+      const currentRenderContext = gridRenderContextSelector(apiRef);
+      const focusedCell = gridFocusCellSelector(apiRef);
+      if (!focusedCell) {
+        return null;
+      }
+
+      const rowIndex = currentPage.rows.findIndex((row) => row.id === focusedCell.id);
+      const columnIndex = visibleColumns.findIndex((column) => column.field === focusedCell.field);
+
+      if (rowIndex === -1 || columnIndex === -1) {
+        return null;
+      }
+
+      const isFocusedCellInContext =
+        currentRenderContext.firstRowIndex <= rowIndex &&
+        rowIndex <= currentRenderContext.lastRowIndex;
+
+      if (isFocusedCellInContext) {
+        return null;
+      }
+      return {
+        ...focusedCell,
+        rowIndex,
+        columnIndex,
+      };
+    },
+    objectShallowCompare,
+  );
```
Another case of "should be a selector with arguments", but this is low-impact, so you could leave a TODO comment and I'm ok with it.

romgrk replied:
Though now that I re-read, those `findIndex()` calls aren't great. Those are `O(n)` costs that are called everytime the state updates, I think it would be preferable to memoize that logic somehow (either through `createSelectorMemoized` or `React.useMemo`).

lauri865 replied:
Good catch. Reduced the lookup cost now, but if you have any ideas on how to make it even better, I'm all ears. Would `rowIdToRowIndex` lookup map make any sense?

lauri865 replied:
Pushed another alternative with lookup maps.

romgrk replied:
I think I don't love the lookup maps, they add an `O(n)` cost (plus a bunch of memory I/O) when the rows are updated to build the lookup map. For the use-cases where rows are updated frequently, it could have an effect. I would favor going back to memoization. If you removed `React.useMemo`, I guess selector memoization would be the alternative. Here is an example how I think it could work (some of the dependency selectors are placeholders, can't remember the right names atm):

https://gist.github.com/romgrk/7f7a682578a1c5f2bbcd8b82aef3148e

lauri865 replied:
As long as there is at least one `O(n)` operation per setting data (which seems to be the case, looking at e.g. `gridExpandedSortedRowEntriesSelector`, but probably not the only one), adding a global map doesn't need to come at an additional cost, and memory aspect can be negated by using the row object reference as key of the map rather than the id. Things could probably be simplified if `useGridSelector` could take non-selector dependencies as well, similar to `useMemo`.

If we can avoid it, I'm all for it, but now when I looked into the implementation, it can become a bit hairy, unless we also refactor `getVisibleRows`. But that we may need to do anyways though, as I just noticed that `useGridVisibleRows` calls an `O(n)` operation on each render by the looks of it, which is a much bigger issue, since the hook is used in 10+ places.

lauri865 replied:
There's currently at least 4 O(n) operations that I found, potentially more, on the example of https://mui.com/x/react-data-grid/#pro-plan:
```
gridSortedRowEntriesSelector 100000
gridExpandedSortedRowEntriesSelector 100000
gridExpandedSortedRowIdsSelector 100000
gridExpandedSortedRowTreeLevelPositionLookupSelector 100000
```
(number shows how many rows the process)

As long as there's at least one of them, adding a lookup map doesn't have to come at an additional cost as long as we can do the operations in parallel. And memory issue can probably be negated by using the row object reference as the map key.

If we can avoid all of them, even better, but maybe not very unlikely? And if so, wouldn't it be better to create the map alongside the other O(n) operations that run an O(n) operation on each focus/rendercontext change?

lauri865 replied:
Seems like there are many places that need to calculate a rowIndex that could potentially be optimised with a lookup map.

lauri865 replied:
`sortedIds` array could be `rowIdToIndex` map instead. Complexity stays the same, but further optimisations possible in the rest of the grid

lauri865 replied:
I don't think I can use `createSelectorMemoized` for this – renderContext will trigger re-evaluation, and we need to re-calculate the `rowIndex`. Unless we start caching focusCell separately in a ref or if we could access previous selector value. However, that makes the code very fragile (what if the sort order changes while the focus cell stays the same? or column order, etc.). Hence I don't currently see a better option than a `lookupMap`, or having `VirtualScroller` re-render on each focus change as it does currently, as then we have the lookup cost only on changing focus, rather than scroll, but that's still less optimal for keyboard navigation in a large grid than a `lookupMap` imho.

The 2nd best option was the first that I submitted – reduce the cost in selector to renderContext size, and calculate real indexes in a `useMemo` for whenever `currentPage.rows` changes.

https://github.com/mui/mui-x/pull/15666/commits/71bffd7ed5b465813f1fbe63b8376e35a61f420a

lauri865 replied:
Pushed a quick example of what I'd do – lookupMap creation takes 10ms in dev for 100k rows under page load (probably less under normal conditions). Would take anywhere from 60-100% longer if we used rowId as the map key instead of object reference, and consume much more memory with uuid ids.

lauri865 replied:
Forgot to update the selector itself, which can now be improved thanks to the change to useGridVisibleRows. But you get the gist.

romgrk replied:
> This selector needs objectShallowCompare still, since it revalidates every time renderContextSelector changes, however, the resulting derived state doesn't always change when renderContextSelector changes, but we can't provide a stable object reference. Would it be different with the v8 selectors?

The trick would be to create intermediate selectors, so that you can compute e.g. `isFocusedCellInContext` when `renderContext` changes, but the value of `isFocusedCellInContext` would stay the same which wouldn't recompute the final `focusedVirtualCell` selector. I think `columnIndex` would also need to come from an intermediate selector.

You can keep the intermediate selectors in the new file, no need to expose every selector publicly.

lauri865 replied:
Would intermediate selectors also have to be memoized or is it enough if the parent selector is? 🤔

romgrk replied:
A selector should be memoized if either of these is true:
 - The selector returns a non-primitive value (object, array, set, map, ...)
 - The selector is computationally expensive (e.g. `O(n)` operation)

So for example the `isFocusedCellInContextSelector` would not need to be memoized, because it returns a boolean which is referentially equal across runs.

romgrk commented on packages/x-data-grid/src/hooks/utils/useGridVisibleRows.ts:
```diff
@@ -36,23 +21,12 @@ export const getVisibleRows = <Api extends GridApiCommon>(
-    }),
-    [response.rows, response.range?.firstRowIndex, response.range?.lastRowIndex],
-  );
+  return useGridSelector(apiRef, gridVisibleRowsSelector, objectShallowCompare);
```
Is `objectShallowCompare` really required here? That thing is a hack, in theory it should be possible to avoid it in all cases that use selectors. The only case where that option makes sense is for inline selectors like `useGridSelector(apiRef, () => { ... })` that can't use memoization.

lauri865 replied:
You're right, it doesn't, removed.

lauri865 commented:
Unit tests are passing now, we'll see about browser ones.

This selector needs `objectShallowCompare` still, since it revalidates every time `renderContextSelector` changes, however, the resulting derived state doesn't always change when `renderContextSelector` changes, but we can't provide a stable object reference. Would it be different with the v8 selectors?

```typescript
const focusedVirtualCell = useGridSelector(
    apiRef,
    gridFocusedVirtualCellSelector,
    objectShallowCompare,
);
```

@romgrk, let me know if you're happy with the change around `useGridVisibleRows` / `getVisibleRows`. If so, we should clean up the current consumers of those methods as well, as they don't need to provide props anymore.

This change forced me to reorder state initialization as well – visible rows are dependent on pagination, which has always been the case, but didn't surface before due to props being passed around.

romgrk commented:
>  let me know if you're happy with the change around useGridVisibleRows / getVisibleRows. 

Very happy, there's so much refactoring that I'd like to do and not enough time. Not sure I'd refactor everything right now, this PR is already large. You can leave a TODO note and/or open a tracking github issue for that.

lauri865 commented:
> > let me know if you're happy with the change around useGridVisibleRows / getVisibleRows.
> 
> Very happy, there's so much refactoring that I'd like to do and not enough time. Not sure I'd refactor every consumer right now, this PR is already large. You can leave a TODO note and/or open a tracking github issue for that.

Makes sense. Already left a to-do note on `useGridVisibleRows` and `getVisibleRows`.

lauri865 commented:
No idea why datepickers related test is failing, but otherwise we're looking good.

lauri865 commented:
@romgrk, just fyi, I'm going on a parental leave beginning of next year, in case you need any additional input from me to get this merged.

The last commits are a bit offtopic, but fixes some issues I hit during rebasing that were introduced by another PR. The issues weren't caught by unit tests due to an extraneous check for virtualization. I only discovered the issues because I needed some props to live within `useGridVirtualScroller`. A fairly small change though and keeps the `renderContext` logic centralised + abolishes the need for any manual overrides to `renderContext` in multiple places. Happy to roll them back if they should slot into another PR though, but that would block this PR until it's merged.

DanailH commented:
@lauri865 thank you for this pull request! Your changes seem to impact the commercially licensed code. For any changes of this nature, we require contributors to sign the MUI’s Contributor License Agreement (CLA). However, I can’t find a CLA signed that could cover these changes. Please follow the steps at https://mui-org.notion.site/CLA-Contributor-License-Agreement-92ece655b1584b10b00e4de9e67eedb0 to sign the document.

PS: For a quicker access you can use this link to fill up the form and effectively "sign" the CLA: https://docs.google.com/forms/d/e/1FAIpQLScPaIH-TzkfyeHqT_QTvwScQ99oujDUh_WWSLeGg747OCVy0w/viewform

romgrk commented:
> @romgrk, just fyi, I'm going on a parental leave beginning of next year, in case you need any additional input from me to get this merged.

Noted, I'm still finishing the review and I can take over the PR if I find anything to change. We only need the CLA signed to allow us to merge it.

lauri865 commented:
CLA signed

DanailH commented:
> CLA signed

Great! Thank you. Can I also ask you to check and do this step https://www.notion.so/mui-org/CLA-Contributor-License-Agreement-92ece655b1584b10b00e4de9e67eedb0?pvs=4#fb0396ecff384a4cb515ab9f4351eddc so we can finish the verification?

lauri865 commented:
Done.

romgrk commented on packages/x-data-grid/src/hooks/features/columnHeaders/useGridColumnHeaders.tsx:
```diff
@@ -159,18 +157,10 @@ export const useGridColumnHeaders = (props: UseGridColumnHeadersProps) => {
-    const { renderContext: currentContext = renderContext, maxLastColumn = visibleColumns.length } =
-      params || {};
-
-    let firstColumnToRender;
-    let lastColumnToRender;
-    if (!rootProps.disableVirtualization && !hasColumnVirtualization) {
-      firstColumnToRender = 0;
-      lastColumnToRender = maxLastColumn;
-    } else {
-      firstColumnToRender = currentContext.firstColumnIndex;
-      lastColumnToRender = currentContext.lastColumnIndex;
-    }
```
I don't see the `rootProps.disableVirtualization` check being present anymore, is this logic gone?

lauri865 replied:
This was recently introduced in the below PR, but is erroneous, as I showcased there – also, I think it's best to avoid these kinds of render context overrides in consumers whenever possible to contain the logic:
https://github.com/mui/mui-x/pull/15116#issuecomment-2555839723

romgrk replied:
Makes sense, I'll be happy to keep that logic in `useGridVirtualScroller`.

romgrk commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -525,13 +502,19 @@ export const useGridVirtualScroller = () => {
         currentRenderContext = frozenContext.current;
       }
 
+      const isVirtualFocusRow = rowIndexInPage === virtualRowIndex;
+      const isVirtualFocusColumn = focusedVirtualCell?.rowIndex === rowIndex;
```
If we're only using the focused-cell state for the virtual logic, then we could eliminate the variable `virtualRowIndex` and use `focusedVirtualCell.rowIndex` directly.

Is the 2nd line correct? Shouldn't `isVirtualFocusColumn` be computed in terms of `columnIndex` instead of `rowIndex`?

lauri865 replied:
That was actually my first thought and implementation as well. But turned out to be incorrect.

Virtual row – row that we mount outside of the render context (=invisible row)
Virtual column – row that is within render context (=visible row), but column is not

`focusedVirtualCell` is defined for both of these cases, but one c

If we would use `focusedVirtualCell.rowIndex` as I did initially, then if we have a row that is within the `renderContext` whose column is outside of the `renderContext`, we would erroneously deem it as `virtualFocusRow`, even though it's actually a visible row. Hope that makes sense.

lauri865 replied:
Actually ´virtualFocusRow' can also have a ´virtualFocusColumn´, since the ´renderContext´ columns are still mounted in an invisible row. It's already handled within the logic, but just missed it in the above description.

Not sure what's the best way to make it more readable though, I did consider some options but didn't land on anything better for now:
* If we get rid of ´virtualRowIndex´, we need to duplicate this code block when iterating over rows: https://github.com/mui/mui-x/blob/57d4f0ed16d5fbca3f583f736f114a068125d222/packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx#L413-L423
* If we also start checking for ´columnIndex´ for ´isVirtualFocusColumn´ in `useGridVirtualScroller´, then we're starting to replicate this code block, and may want to pass it as props, but feels like it belongs best to `GridRow`: https://github.com/mui/mui-x/blob/85961a5832af9c358b67ed62e918dbbf7817e205/packages/x-data-grid/src/components/GridRow.tsx#L117-L125
* Alternatively, we could leave some comments

The current approach felt like the least error prone to me.

romgrk replied:
Makes sense, and the virtual focus logic is working as expected, so let's keep it as it is.

romgrk commented on packages/x-data-grid/src/hooks/utils/useGridVisibleRows.ts:
```diff
@@ -1,32 +1,16 @@
 export const getVisibleRows = <Api extends GridApiCommon>(
   apiRef: React.MutableRefObject<Api>,
-  props: Pick<DataGridProcessedProps, 'pagination' | 'paginationMode'>,
+  // To-do: remove after getVisibleRows implementations have been updated
```
nitpick: here and below, `s/To-do/TODO/`.

romgrk commented on packages/x-data-grid/src/hooks/features/pagination/gridPaginationSelector.ts:
```diff
@@ -16,6 +17,15 @@ const ALL_RESULTS_PAGE_VALUE = -1;
  */
 export const gridPaginationSelector = (state: GridStateCommunity) => state.pagination;
 
+/**
+ * @category Pagination
+ * @ignore - do not document.
+ */
+export const gridPaginationEnabledClientSideSelector = createSelector(
```
This is being exported from the root `index.ts` despite being `@ignore`, it would be nice to avoid exposing it.

romgrk commented:
The merge commit was painful, waiting to see what the CI says but I might have made errors on that one.

lauri865 commented:
@romgrk, squashed and merged it. Unit tests are passing locally at least, hope the rest of the CI will be happy as well. Please don't make me do this again 🙈.

lauri865 commented:
Fixed a test that started failing. I'm not sure why it works even currently, as it's testing pagination where pagination should be disabled in the Pro version by default. As per docs: ´For the Pro and Premium Data Grid, pagination is disabled by default; use the pagination prop to enable it.´

Added the necessary props, and it's passing locally now.

romgrk commented:
> @romgrk, squashed and rebased it. Unit tests are passing locally at least, hope the rest of the CI will be happy as well. Please don't make me do this again 🙈.

Really sorry about that, I also tried to merge & fix conflicts so I know how painful it must have been. Sidenote but I'm curious why you went for a rebase, I find fixing conflicts much harder for a rebase than for a merge.

lauri865 commented:
> > @romgrk, squashed and rebased it. Unit tests are passing locally at least, hope the rest of the CI will be happy as well. Please don't make me do this again 🙈.
> 
> Really sorry about that, I also tried to merge & fix conflicts so I know how painful it must have been. Sidenote but I'm curious why you went for a rebase, I find fixing conflicts much harder for a rebase than for a merge.

I agree, but with `git merge --squash` I still ended up having to replay the merge conflict history to a degree. Technically it was still a merge in the end. I squashed (that was the rebase) and merged in one go, to avoid having to replay the whole merge conflict history. Not sure if there's a better way or if my git knowledge is failing me, but it does the trick for me at least when things get messy.

romgrk reviewed (approved):
Waiting for the CI to pass for the small refactor commit I added, but other than that this is ready to be merged.
